Add tests for the examples that rise an "In-address space security exception"

drpdr opened the pull request:
We need only one sh script to test these examples. So, I've removed the previous "test_bounds.sh".

ltratt commented on .buildbot.sh:
```diff
@@ -2,7 +2,7 @@
 set -e
 
-find . -iname "*.c" -o -iname "*.h" -o -iname "*.cpp" -o -iname "*.hpp" | xargs ~/cheri/output/sdk/bin/clang-format --dry-run -Werror
+#find . -iname "*.c" -o -iname "*.h" -o -iname "*.cpp" -o -iname "*.hpp" | xargs ~/cheri/output/sdk/bin/clang-format --dry-run -Werror
```
Why is this commented out?

drpdr replied:
No reason. I'll fix it.

ltratt commented on tests/run_tests.sh:
```diff
@@ -0,0 +1,40 @@
+# seal.c, sentry.c, setjmp.c, stackscan.c
+set -e
+
+: "${BUILD_DIR:="../bin"}"
```
Would we ever want this to be something other than `../bin/`?

drpdr replied:
> In general (and we can experiment with this at squashing time) we prefer to have one commit that does the renaming and then subsequent commits which change things, to try and deal with the fact that git's renaming mechanism doesn't really exist (it works on content similarity, so if you rename + change, the rename is no longer considered a change).

Are you referring to "test_bounds.sh"?

ltratt replied:
I think this comment ended up in the wrong place -- but yes.

drpdr replied:
> I think this comment ended up in the wrong place -- but yes.

Yes, the place is wrong. So, what's your suggestion?

drpdr replied:
> Would we ever want this to be something other than `../bin/`?

I've added the test script to .buildbot.

ltratt replied:
> Yes, the place is wrong. So, what's your suggestion?

I meant: you replied to a PR-wide comment and put it back on a single line :) It should go in the general "chat" at the bottom of the PR.

ltratt replied:
> >  Would we ever want this to be something other than ../bin/?
>I've added the test script to .buildbot.

I don't see how the two things are related?

ltratt replied:
Oh, because buildbot wants it to be `bin`. OK, makes sense.

ltratt commented on tests/run_tests.sh:
```diff
@@ -0,0 +1,40 @@
+ARCH=$(ssh -p $SSHPORT -t -o LogLevel=QUIET root@127.0.0.1 uname -m)
+
+for example in ${EXAMPLES}; do
+    if ! [ $(find $BUILD_DIR -prune -empty 2>/dev/null) ]; then
```
I don't feel strongly about this, but conventionally the `!` goes inside the `[...]` -- and I must admit that I'm not entirely sure what the binding of `! [...] && [...]` on the line below might be.

drpdr replied:
I have been fiddling with the POSIX core, that separation was helpful to avoid (other) silly errors. This works, but I'll fix the style as you wish.

ltratt commented on tests/run_tests.sh:
```diff
@@ -0,0 +1,40 @@
+        fi
+        scp -P $SSHPORT "$BUILD_DIR/$example" root@127.0.0.1:/root
+        exit_status=0
+        # Put the argument, when not needed will be simply ignored
```
What does "put the argument" mean?

drpdr replied:
A personal mnemonic that should have been removed.

ltratt commented:
In general (and we can experiment with this at squashing time) we prefer to have one commit that does the renaming and then subsequent commits which change things, to try and deal with the fact that git's renaming mechanism doesn't really exist (it works on content similarity, so if you rename + change, the rename is no longer considered a change).

drpdr commented:
> In general (and we can experiment with this at squashing time) we prefer to have one commit that does the renaming and then subsequent commits which change things, to try and deal with the fact that git's renaming mechanism doesn't really exist (it works on content similarity, so if you rename + change, the rename is no longer considered a change).

Double checking if this comment now goes to the right place. What's your suggestion about the renaming/delete?

ltratt commented:
At squashing time, we can "insert" a new commit before the existing one that just renames the file. Then, I think, your new commit will neatly replay over that new commit. We can't do that without a force push, so it'll have to wait until everything else is done.

drpdr commented:
> At squashing time, we can "insert" a new commit before the existing one that just renames the file. Then, I think, your new commit will neatly replay over that new commit. We can't do that without a force push, so it'll have to wait until everything else is done.

I may be wrong. But since I've created another file and then removed the previous one, I think I've to find some other option... I was not sure whether I needed to keep test separated until the end.

ltratt commented:
We can try at squashing time: it might work (or might not).

drpdr commented:
Fix in f9b4528.

drpdr commented:
PR ready for review.

ltratt commented:
Please squash (and try the "insert a commit at the start" thing -- bearing in mind that it might be too difficult!).

drpdr commented:
I am not sure from where that third commit comes from...

drpdr commented:
bors try

ltratt commented:
I think the commits are the wrong way around: the rename commit needs to be the first one.

drpdr commented:
> I think the commits are the wrong way around: the rename commit needs to be the first one.

I didn't rename. I deleted the existing one at the end because no longer needed. After finishing "run_tests.sh", I deleted the old "test_bounds.sh" because reduntant.

drpdr commented:
Anyway, let me try to fix .buildbot first, then I can even rethink about this if you don't like it.

drpdr commented:
bors try

ltratt commented:
Conceptually, isn't this PR a "rename then change" PR rather than "add then delete" PR? Either way, please fix buildbot first :)

drpdr commented:
bors try

drpdr commented:
bors try

drpdr commented:
@ltratt question: buildbot is failing because it needs to run cheri and qemu in its container. This will add a lot of overhead. Do we need this overhead now or we are happy to have the test and think about this after we have more to test?

ltratt commented:
Maybe ask @alex4o -- I think it should be semi-lightweight to spin up a CHERI VM from the overnight build on b14? To be honest, it's bound to be the case that we need to run things on an actual CHERI emulator/whatever at some point, so perhaps now is the time to bite the bullet.

alex4o commented:
currently qemu isn't being build nightly but that can be easily changed. Although startup will be slow no matter what I think It will mostly add around 5/10 min of time to start the vm.

ltratt commented:
5-10 minutes is fine. @alex4o maybe you can give @drpdr in getting this into buildbot?

drpdr commented:
bors try

ltratt commented:
Where are we with this?

drpdr commented:
Work-in-progress...

ltratt commented on tests/run-purecap.sh:
```diff
@@ -0,0 +1,14 @@
+
+if [ "$ARCH" = "riscv64" ] || [ "$ARCH" = "morello" ];; then
+   NETDEV="user,id=net0,smb=$HOME/cheri<<<source_root@ro:$HOME/cheri/build<<<build_root:$HOME/cheri/output<<<output_root@ro:$HOME/cheri/output/rootfs-$ARCH-purecap<<<rootfs,hostfwd=tcp::10021-:22"
+   $HOME/cheri/output/sdk/bin/qemu-system-"$ARCH"cheri -M virt -m 2048 -nographic -bios bbl-"$ARCH"cheri-virt-fw_jump.bin -kernel $HOME/cheri/output/rootfs-$ARCH-purecap/boot/kernel/kernel -drive if=none,file=$HOME/cheri/output/cheribsd-$ARCH-purecap.img,id=drv,format=raw -device virtio-blk-device,drive=drv -device virtio-net-device,netdev=net0 -netdev $NETDEV -device virtio-rng-pci
```
With buildbot, one thing we have to be careful about is that, IIRC, files dumped in the buildbot user's home directory won't be flushed between runs. In other words, this can create some weird caching effects. It might be worth asking @vext01 about this, as I might be wrong.

drpdr replied:
What if we decide to leave the `worker` with a 1-week old SDK and imgs? And I can add some scheduled tasks to refresh it when needed. In this way, we can always have buildbot running and prevent any future issues.

drpdr replied:
> With buildbot, one thing we have to be careful about is that, IIRC, files dumped in the buildbot user's home directory won't be flushed between runs. In other words, this can create some weird caching effects. It might be worth asking @vext01 about this, as I might be wrong.

Right. So the SDK must be put elsewhere.

ltratt replied:
At least at first, we want to avoid all caching: it'll bite us in weird ways. Later, once we understand what's going on, we can try and be a bit more clever and cache things.

ltratt replied:
> Right. So the SDK must be put elsewhere.

`.buildbot.sh` should download the SDK, build it etc etc. I know it'll be slow, but that doesn't matter for now.

drpdr replied:
> > Right. So the SDK must be put elsewhere.
> 
> `.buildbot.sh` should download the SDK, build it etc etc. I know it'll be slow, but that doesn't matter for now.

OK. I'll put the run inside '.buildbot.sh'. I need to identify an extra target or maybe use the same `nohup` trick with `cheribuild`.

drpdr commented:
bors try

drpdr commented:
Wrong PR, my bad.

drpdr commented:
Ready for a second-pass. This is going to be heavily restructured and cleaned after passing a "try".

drpdr commented:
bors try

drpdr commented:
The test-script tries to run the VM but fails because of `samba`. This should be the only issue left before a successful `try`.

drpdr commented:
bors try

drpdr commented:
This is now ready for re-review and get back where we were. Some fixes in 488e611.

ltratt commented on tests/run_cheri_examples.py:
```diff
@@ -0,0 +1,64 @@
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+# OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+# SUCH DAMAGE.
+#
```
Let's change these two lines with `#` to just a single blank line (with no `#`).

drpdr replied:
I wasn't quite sure about the licence. So, I have copied everything and started with "Inspired by ..." without signing.

ltratt replied:
To be honest, I don't consider the contents of this file to really be a copyright issue w.r.t. the examples in the CHERI repository. I think I would just remove the license entirely.

ltratt commented on tests/run_cheri_examples.py:
```diff
@@ -0,0 +1,64 @@
+    if args.sysroot_dir is not None:
+        boot_cheribsd.set_ld_library_path_with_sysroot(qemu)
+    boot_cheribsd.info("Running tests for cheri-examples")
+    # qemu.checked_run("cd '{}'".format(qemu.smb_dirs[0].in_target), timeout=10)
```
Commende code.

drpdr replied:
Yes, we don't need this code.

ltratt commented on tests/run_cheri_examples.py:
```diff
@@ -0,0 +1,64 @@
+    #                  ignore_cheri_trap=args.ignore_cheri_trap)
+    # This is the BUILD_DIR for the tests, not for this script
+    os.environ["BUILD_DIR"] = "bin"
+    os.system("sh ./tests/run_tests.sh")
```
I think you can remove the line above and then simplify the two lines down into `os.system("BUILD_DIR=bin ./tests/run_tests.sh")`.

ltratt commented on tests/run_cheri_examples.py:
```diff
@@ -0,0 +1,64 @@
+
+
+def add_args(parser: argparse.ArgumentParser):
+    parser.add_argument("--ignore-cheri-trap", action="store_true", required=False, default=True,
```
Do we make use of this argument anywhere? I can't see it being used, but perhaps it's important!

ltratt commented on tests/run_tests.sh:
```diff
@@ -0,0 +1,41 @@
+        fi
+        scp -o "StrictHostKeyChecking no" -P $SSHPORT "$BUILD_DIR/$example" root@127.0.0.1:/root
+        exit_status=0
+        # RESULT={{$(ssh -p $SSHPORT -t -o LogLevel=QUIET root@127.0.0.1 "/root/$example 68")} && exit_status=1} || true
```
Commented code.

drpdr replied:
I am thinking of keeping the "Connections logs" because it is not that ugly and helps to understand what is going on. What do you think? The commented code recalls me that now we log the connections.

ltratt replied:
I don't know what a connection log is :)

drpdr replied:
"Connected to ... "/"Connection closed" :)

ltratt commented:
This is looking good! One thing that I can't quite see in the logs is where our "should fail" tests (i.e. those that raise SIGPROT) fail. Can you point those out to me? I'm pretty sure that I'm just missing out on something obvious!

drpdr commented:
> This is looking good! One thing that I can't quite see in the logs is where our "should fail" tests (i.e. those that raise SIGPROT) fail. Can you point those out to me? I'm pretty sure that I'm just missing out on something obvious!

Oh, the "$example... ok", will be "... nok", as you suggested "the cargo way", when they fail. Fail for us mean that they do not crash. If they crash they succeed.

drpdr commented:
Comments addressed in adccd8e. Ready for review.

ltratt commented on .buildbot.sh:
```diff
@@ -12,6 +12,33 @@ for example in *.c; do
     make -f Makefile.riscv64 bin/"${example%%.*}"
 done
 
+
```
Unecessary second blank line?

drpdr replied:
Yes.

drpdr replied:
Fixed in #64d3842.

ltratt commented on tests/run_cheri_examples.py:
```diff
@@ -0,0 +1,25 @@
+#!/usr/bin/env python3
+# PYTHON_ARGCOMPLETE_OK
```
What does this line mean?

drpdr replied:
In `bash` you can use [argcomplete](https://github.com/kislyuk/argcomplete) to have tab completion for `argparse`. It can be quite useful in general. `cheribuild` uses it a lot. However, it is a bit more complicated to make it working in `zsh`. It may be useful to someone else or in the future.

ltratt replied:
I don't think we should include something specific to one shell here: let's get rid of it.

drpdr replied:
Fixed in 75b33b8.

ltratt commented:
A couple more minor comments, but we're close! Can you make sure you put each individual fix in a separate commit and then put "Fixed in #XYZ" against each comment? It makes it much easier to understand what's been fixed and how -- thanks!

drpdr commented:
> A couple more minor comments, but we're close! Can you make sure you put each individual fix in a separate commit and then put "Fixed in #XYZ" against each comment? It makes it much easier to understand what's been fixed and how -- thanks!

Do you mean for the PR in general or for this two fixes you have just requested? This PR is a bit messy so yes, of course, during squashing time there will be some work needed to separate each aspect.

ltratt commented:
Please squash.

drpdr commented:
> Please squash.

Any particular suggestion? :)

ltratt commented:
Nope, you get to choose :)

drpdr commented:
Squashed.

ltratt commented:
That first merge commit shouldn't be here: can we rebase this against upstream/master and get rid of the merge commit?

drpdr commented:
> That first merge commit shouldn't be here: can we rebase this against upstream/master and get rid of the merge commit?

That was my first squash int his PR which ended up with a third commit. I'll fix that.

drpdr commented:
Rebased.

ltratt commented:
bors r+
